### PR TITLE
Bug 2034563: validate azure ephemeral storage location and caching type

### PR DESCRIPTION
**Pull request by Fedosin**
This commit adds missing validations for the recently added osDisk properties: diskSettings.ephemeralStorageLocation and cachingType [1]

[1] https://github.com/openshift/api/pull/1054

### CODE REVIEW

File: pkg/webhooks/machine_webhook.go [hunk 1]
```diff
@@ -85,7 +85,13 @@ const (
...
 	defaultAzureOSDiskOSType      = "Linux"
 	defaultAzureOSDiskStorageType = "Premium_LRS"
-	azureMaxDiskSizeGB            = 32768
+
+	// Azure OSDisk constants
+	azureMaxDiskSizeGB                 = 32768
+	azureEphemeralStorageLocationLocal = "Local"
+	azureCachingTypeNone               = "None"
+	azureCachingTypeReadOnly           = "ReadOnly"
+	azureCachingTypeReadWrite          = "ReadWrite"
```
Comment (JoelSpeed) [inline pkg/webhooks/machine_webhook.go:94]: Are there constants in the openshift/api we can use here instead?

File: pkg/webhooks/machine_webhook.go [hunk 2]
```diff
@@ -859,6 +865,21 @@ func validateAzure(m *machinev1.Machine, config *admissionConfig) (bool, []strin
 		errs = append(errs, field.Invalid(field.NewPath("providerSpec", "osDisk", "diskSizeGB"), providerSpec.OSDisk.DiskSizeGB, "diskSizeGB must be greater than zero and less than 32768"))
 	}
 
+	if providerSpec.OSDisk.DiskSettings.EphemeralStorageLocation != azureEphemeralStorageLocationLocal && providerSpec.OSDisk.DiskSettings.EphemeralStorageLocation != "" {
+		errs = append(errs, field.Invalid(field.NewPath("providerSpec", "osDisk", "diskSettings", "ephemeralStorageLocation"), providerSpec.OSDisk.DiskSettings.EphemeralStorageLocation,
+			fmt.Sprintf("osDisk.diskSettings.ephemeralStorageLocation can be either empty or set to %s", azureEphemeralStorageLocationLocal)))
```
Comment (JoelSpeed) [inline pkg/webhooks/machine_webhook.go:870]: This wording update makes it fit the same style as the comments in the openshift API repo
```suggestion
			fmt.Sprintf("osDisk.diskSettings.ephemeralStorageLocation can either be omitted or set to %s", azureEphemeralStorageLocationLocal)))
```

File: pkg/webhooks/machine_webhook.go [hunk 3]
```diff
@@ -859,6 +865,21 @@ func validateAzure(m *machinev1.Machine, config *admissionConfig) (bool, []strin
...
+	}
+
+	cachingTypes := []string{azureCachingTypeNone, azureCachingTypeReadOnly, azureCachingTypeReadWrite}
+	if !stringInSlice(providerSpec.OSDisk.CachingType, cachingTypes) && providerSpec.OSDisk.CachingType != "" {
+		errs = append(errs, field.Invalid(field.NewPath("providerSpec", "osDisk", "cachingType"), providerSpec.OSDisk.CachingType,
+			fmt.Sprintf("osDisk.cachingType can be only %s or an empty string", strings.Join(cachingTypes, ", "))))
+	}
```
Comment (JoelSpeed) [inline pkg/webhooks/machine_webhook.go:877]: Alternatively you can use a switch here, and you can avoid the list iteration `stringInSlice`
```suggestion
	switch providerSpec.OSDisk.CachingType:
	case azureCachingTypeNone, azureCachingTypeReadOnly, azureCachingTypeReadWrite, "":
	  // Valid scenarios, do nothing
	default:
	  errs = append(errs, field.Invalid(field.NewPath("providerSpec", "osDisk", "cachingType"), providerSpec.OSDisk.CachingType,
			fmt.Sprintf("osDisk.cachingType can be only %s, %s, %s or omitted", azureCachingTypeNone, azureCachingTypeReadOnly, azureCachingTypeReadWrite)))	
	}
```

File: pkg/webhooks/machine_webhook.go (hunk 1 above)
Comment (Fedosin) [reply pkg/webhooks/machine_webhook.go:94]: These constants are defined in the Azure SDK, which is not in the dependencies. Probably it doesn't make much sense to import it in this case.

**Review by JoelSpeed — COMMENTED** [review]
/lgtm

**Pull request merged**